### PR TITLE
Handle UPD 15,22,27

### PR DESCRIPTION
**Pull request by samuel-williams-da**
- [aead2b8](https://github.com/digital-asset/docs.daml.com/pull/874/commits/aead2b8955240aefa14620a9c241f4d66ab385e5): UPD 27, module prefixes.
- [a03a60a](https://github.com/digital-asset/docs.daml.com/pull/874/commits/a03a60aeb77b1f70e850d92a004be2bdc0167753): UPD 15 to the best of my ability, verify limitations
- [8bf9deb](https://github.com/digital-asset/docs.daml.com/pull/874/commits/8bf9debe9e397a22e480c04ebb89e74c56855e3d): UPD 22, daml script testing recommendation
- [b3c49a6](https://github.com/digital-asset/docs.daml.com/pull/874/commits/b3c49a61edd01265d7641f1140eb3e235676c5fd): Additional point about interface instance changes, related to UPD 15

To be merged after https://github.com/digital-asset/docs.daml.com/pull/871

### CODE REVIEW

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 1]
```diff
@@ -1948,11 +1947,8 @@ it as two packages, ``helper`` and ``main``:
 Remove Retroactive Instances
 ----------------------------
 
-SCU eliminates the need for retroactive instances and is not
-compatible with them. To ensure package selection for interface choices
-acts correctly, retroactive interface instances should be moved to newer
-versions of templates, such that changes to the instance warrants a new
-version of the template.
+SCU is not compatible with retroactive instances, so these will need to be removed.
```
Comment (paulbrauner-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:1950]: In which situation would we "remove" them rather than not introduce them? I guess if you're compiling some old code to LF 1.17?

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 2]
```diff
@@ -2082,7 +2078,60 @@ Daml Script has been used for demonstrative purposes in this document, however
 usually the complexities of live upgrades comes with your workflows, not the data
 transformations themselves. You can use Daml Script (with Canton) to test some
 isolated simple cases, but for thorough tests of you system using SCU, you should
-prefer full workflow testing, as below.
+prefer full `Workflow Testing <#workflow-testing>`__.
+
+When testing your upgrades with Daml Script, we recommend you place your Daml Script tests
+in a separate package which depends on all versions of your business logic. This testing
+package should not be uploaded to the ledger if possible, as it cannot be dropped in an upgrade.
```
Comment (paulbrauner-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2085]: "as it cannot be dropped in an upgrade": technically you can no? Since it's a separate package that presumably no other package depends on, you can choose not to upload any upgrade to it. You can even unvet it.

It seems to me the real reason we don't want users to upload such a package is because it depends on daml-script and we don't want users to upload daml-script-lts.

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 3]
```diff
@@ -2082,7 +2078,60 @@ Daml Script has been used for demonstrative purposes in this document, however
...
+   | These new data types should mostly just hold Contract IDs
+
+-  | Update your existing test suite from V1 to take a :ref:`package preference <daml-script-package-preference>`,
+     allowing the V2 implementation without additional fields to choices to be called.
+   | ``testV1 : [PackageId] -> V1TestData -> Script ()``
+
+-  | Run the above test suite against V1 data, passing a V1 preference, then a V2 preference.
+   | This ensures your changes haven't broken any existing workflows.
+
+-  | Next write tests for your new/modified workflows, using the V2 choice implementations. This doesn't not need a package preference.
```
Comment (paulbrauner-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2125]:
```suggestion
-  | Next write tests for your new/modified workflows, using the V2 choice implementations. This doesn't need a package preference.
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 4]
```diff
@@ -2082,7 +2078,60 @@ Daml Script has been used for demonstrative purposes in this document, however
...
+   | ``testV2 : TestData -> Script ()``
```
Comment (paulbrauner-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2126]: Should that be V2TestData?

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 5]
```diff
@@ -2082,7 +2078,60 @@ Daml Script has been used for demonstrative purposes in this document, however
...
+
+-  | Run these tests against both the V1 setup and the V2 setup, to ensure your new workflows support existing/old templates.
+   | In order to do this, you'll need some way to upcast your ``V1TestData``, i.e.
+   | ``upcastTestData : V1TestData -> V2TestData``
+   | This function should mostly just call ``coerceContractId`` on any contract IDs, and fill in any ``None`` values if needed.
+
+-  | Finally, you can cover any workflows that require the contract data to already be upgraded (new fields populated), these are
```
Comment (paulbrauner-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2133]: Not an english expert but is that correct? Shouldn't that either be "any workflow that requires" or "worflows that require"?

**Review by paulbrauner-da — CHANGES_REQUESTED** [review]
Looks good! Left a few questions.

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 5 above)
Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2133]: I think this is fine, for example you can say like "Do you have any ideas". Its okay to follow `any` by a plural

Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2133]: I could reword it to `all workflows`, I'll wait to see what the docs team thinks

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 6]
```diff
@@ -365,18 +367,15 @@ In this version of SCU, the following functionality has not yet
 been implemented, but may be implemented in future releases.
 
 -  Retroactive interface instances are not compatible with SCU upgrades.
-   We do not expect to ever add this compatibility, as SCU supersedes retroactive
-   interface instances
+   SCU partially covers this behaviour, allowing instances to be changed in an upgrade,
```
[the remark this answers is not in the document]
Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:370]: Fair enough, I'll update that

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 1 above)
Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:1950]: Yeah, this is more in the context of migration, but migration points to this list. I'll see if theres better wording

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 5 above)
Comment (paulbrauner-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2133]: If it sounds good to you, an english native speaker, then I think that's ok :)

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 2 above)
Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2085]: This is true, but given unvetting isn't deletion, they would still be stuck with it on their ledger.
I'll try a better wording, but i really do mean in general, dont upload daml script

Comment (paulbrauner-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2085]: Oh absolutely! My point was maybe what we should say just that: don't upload that package because it will upload daml-script and we don't want that.

**Review by paulbrauner-da — APPROVED** [review]
Thank you!

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 7]
```diff
@@ -355,6 +355,8 @@ data transformations that cannot be made using SCU upgrades:
 
 -  Upgrading interface and exception definitions
 
+-  Adding/removing an interface instance on a template
```
Comment (dylant-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:358]: :+1:

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 8]
```diff
@@ -1948,10 +1947,10 @@ it as two packages, ``helper`` and ``main``:
 Remove Retroactive Instances
 ----------------------------
 
-SCU eliminates the need for retroactive instances and is not
-compatible with them. Retroactive interface instances should be moved to newer
-versions of templates, such that changes to the instance warrant a new
-version of the template, to ensure that the correct package is selected for interface choices.
+SCU is not compatible with retroactive instances, so if you are migrating from an LF1.15
+project that uses this, you'll need to move the instances to their respective templates
```
Comment (dylant-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:1951]:
```suggestion
SCU is not compatible with retroactive interface instances, so if you are migrating to SCU from an LF1.15
project that uses retroactive instances, you'll need to move the instances to their respective templates
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 9]
```diff
@@ -2078,7 +2077,62 @@ Daml Script has been used for demonstrative purposes in this document, however
...
 transformations themselves. You can use Daml Script (with Canton) to test some
 isolated simple cases, but for thorough tests of you system using SCU, you should
-prefer full workflow testing, as below.
+prefer full `Workflow Testing <#workflow-testing>`__.
+
+When testing your upgrades with Daml Script, we recommend you place your Daml Script tests
+in a separate package which depends on all versions of your business logic. This testing
+package should not be uploaded to the ledger if possible, as it depends on the ``daml-script-lts`` package.
+This package emits a warning on the participant when uploaded, as it serves no purpose on a participant,
+cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3).
```
Comment (dylant-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2086]:
```suggestion
cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3). More information about this limitation :ref:`here <upgrade_dont_upload_daml_script>`.
```

**Review by dylant-da — APPROVED** [review]
LGTM, I like your fixes to Paul's requested changes.

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 5 above)
Comment (dylant-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2134]: I think it's OK, there's a small but in this case meaningless difference I think.

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 9 above)
Comment (dylant-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2086]: This reference will be there once https://github.com/digital-asset/docs.daml.com/pull/878 goes in

Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2086]: This doesn't exist yet, the :ref: would error on build

Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2086]: I suspect this PR will merge first, given it already has language team reviews. You may want to update this in your PR.
We'll see, race condition

Comment (dylant-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2086]: Yeah I figured - left it here not to forget

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 10]
```diff
@@ -1948,10 +1947,10 @@ it as two packages, ``helper`` and ``main``:
 Remove Retroactive Instances
 ----------------------------
 
-SCU eliminates the need for retroactive instances and is not
-compatible with them. Retroactive interface instances should be moved to newer
-versions of templates, such that changes to the instance warrant a new
-version of the template, to ensure that the correct package is selected for interface choices.
+SCU is not compatible with retroactive interface instances, so if you are migrating to SCU from an LF1.15
+project that uses retroactive instances, you'll need to move the instances to their respective templates
+during the migration.
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:1952]:
```suggestion
project that uses retroactive instances, you must move the instances to their respective templates
during the migration.
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 9 above)
Comment (carrielaben-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2086]:
```suggestion
We recommend placing your Daml Script tests
in a separate package which depends on all versions of your business logic when testing your upgrades with Daml Script. This testing
package should not be uploaded to the ledger if possible, as it depends on the ``daml-script-lts`` package.
This package emits a warning on the participant when uploaded, as it serves no purpose on a participant,
cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3).
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 11]
```diff
@@ -2078,7 +2077,62 @@ Daml Script has been used for demonstrative purposes in this document, however
...
-prefer full workflow testing, as below.
+prefer full `Workflow Testing <#workflow-testing>`__.
+
+When testing your upgrades with Daml Script, we recommend you place your Daml Script tests
+in a separate package which depends on all versions of your business logic. This testing
+package should not be uploaded to the ledger if possible, as it depends on the ``daml-script-lts`` package.
+This package emits a warning on the participant when uploaded, as it serves no purpose on a participant,
+cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3).
+Depending on multiple versions of the same package will however face ambiguity issues with
+imports, you can resolve this using :ref:`module prefixes <module_prefixes>`:
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2088]:
```suggestion
Depending on multiple versions of the same package does however face ambiguity issues with
imports. You can resolve these issues using :ref:`module prefixes <module_prefixes>`:
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 12]
```diff
@@ -2078,7 +2077,62 @@ Daml Script has been used for demonstrative purposes in this document, however
...
+    - my/v1/main-package.dar
+    - my/v2/main-package.dar
+  module-prefixes:
+    main-1.0.0.dar: V11
+    main-1.1.0.dar: V12
+
+For writing your tests, it is important that you verify old workflows are still functional under
+new data, and new implementation. You'll also need to verify that new workflows (that are intended
+to be backwards compatible) are able to consume old data. You should build your testing structure to
+cover this how you see fit, but we give the following recommendation as a starting point:
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2104]:
```suggestion
It is important to verify old workflows are still functional under
new data and new implementation when writing your tests. You also need to verify that any new workflows intended
to be backward compatible can consume old data. You should build your testing structure to
cover this how you see fit, but we give the following recommendation as a starting point:
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 13]
```diff
@@ -2078,7 +2077,62 @@ Daml Script has been used for demonstrative purposes in this document, however
...
+
+If your new version only includes choice body or interface instance changes (i.e. it is a patch release)
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2106]:
```suggestion
If your new version only includes choice body or interface instance changes (that is, it is a patch release)
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 14]
```diff
@@ -2078,7 +2077,62 @@ Daml Script has been used for demonstrative purposes in this document, however
...
+
+-  | Simply run your existing test suite written for V1 but updated to call V2 choices. This can be
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2108]:
```suggestion
-  | Run your existing test suite for V1 but updated to call V2 choices. This can be
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst [hunk 15]
```diff
@@ -2078,7 +2077,62 @@ Daml Script has been used for demonstrative purposes in this document, however
...
+     done with a rewrite, or by passing down a :ref:`package preference <daml-script-package-preference>`
+     and calling the test with both the V1 and V2 package ID.
+
+If your new version includes data changes, be that to contract payloads or choices (i.e. it is a minor release)
```
Comment (carrielaben-da) [inline docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2112]:
```suggestion
If your new version includes data changes, be that to contract payloads or choices (that is, it is a minor release)
```

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 5 above)
Comment (carrielaben-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2134]: "any workflows" and "all workflows" are equally correct in this case, since "any" implies it could be one or many.

**Review by carrielaben-da — APPROVED** [review]
Small changes.

File: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst (hunk 9 above)
Comment (samuel-williams-da) [reply docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst:2085]: @dylant-da This PR has merged, you'll need to make this addition on your side
